### PR TITLE
CIP-0115? | CBOR tag definition - ED25519-BIP32 Keys

### PR DESCRIPTION
**Pull request by stevenj**
This CIP is both desired by Project Catalyst to help define CBOR data structures, and serves also as an example implementation for https://github.com/cardano-foundation/CIPs/pull/752.

Note, the example registry in https://github.com/cardano-foundation/CIPs/pull/752 has matching definitions for these tags, typically a PR would include changes for that file once its been merged.

This is the third of three PRs that should be considered together for context.

See also:

* #751 
* #752 

Closes input-output-hk/catalyst-CIPs#4

---

[Rendered proposal on author's fork](https://github.com/input-output-hk/catalyst-CIPs/tree/cip/ed25519-tags/CIP-0115)

### CODE REVIEW

**Comment by Ryun1** [comment]
Should this be its own CIP or could this just be added as a document inside of https://github.com/cardano-foundation/CIPs/pull/752?

Just thinking if it is easier to manage if contained in one CIP?

**Comment by stevenj** [comment]
> Should this be its own CIP or could this just be added as a document inside of #752?
> 
> Just thinking if it is easier to manage if contained in one CIP?

I didn't think it was normal to keep editing CIPs once they are accepted.
There are a bunch of candidates for tags, this is just a really easy one I need.
CIP-19 jumpts to mind as well.
Following the example of the metadata registry CIP it just explains how, and new metadata CIPs are supposed to include a patch to its registry file.  All metadata isn't defined in CIP-10

**Comment by rphair** [comment]
@Ryun1 @stevenj adding CIPs for tag definitions would be consistent with our precedent for CIP-0068 - for which the metadata and motivation / rationale would be hopelessly more complicated than we could ever simply tabulate in the originally conceived CIP-0067 table to store those asset labels.

Since we've told new asset label (token type) creators to submit new CIPs for any new definitions beyond the first four, it would make sense to do so here... and even more sense if @stevenj your planned introduction of these key/tag definitions can be broken into groups as large as possible, as you already appear to have done here.

On the other end of the complexity spectrum: we were able to "get away with" using a simple JSON array in CIP-0010 because the information stored was simpler, and the details of each entry were colloquial and could be challenged in the PR if necessary (cc @KtorZ).  But the information associated with each of these CBOR tag/keys needs more authentication that needs to be recorded more officially: so requiring new CIPs does make sense.

@stevenj one thing we need to stipulate for both CIP-0013 and CIP-0068 (cc @Crypto2099)... to make both of them "extensible"... is to say very clearly that a certain number of initial definitions were supported in the original CIP and then others beyond would follow a specific form.  You may already have done this but please keep this "plug-in" convenience as a goal when refining your draft for #752.

... and therefore @Ryun1 that would indeed allow a number of initial definitions to included as suggested in https://github.com/cardano-foundation/CIPs/pull/753#issuecomment-1911992973.

**Comment by rphair** [comment]
@stevenj @Ryun1 - separate question from the last: I'm avoiding tagging this `Metadata` or `Tools` until we settle the question asked earlier in https://github.com/cardano-foundation/CIPs/pull/752#discussion_r1466557385.

**Comment by Ryun1** [comment]
Agreed @stevenj @rphair 

lets do new CIPs!

**Comment by rphair** [comment]
@Ryun1 it does seem like it would be cleaner to have 1 CIP for the framework and 1 CIP for each new set of related keys.  We were kind of stuck with CIP-0068 because we had to represent the state of the art until that point... but here we can start fresh, so there's no reason the framework CIP would have to include a "base" set of keys (I'd be happy to leave that to @stevenj this time).

This can be confirmed during Triage at the next meeting where it would be up for introduction anyway (I've put it on the agenda): https://hackmd.io/@cip-editors/81

**Comment by Ryun1** [comment]
> @Ryun1 it does seem like it would be cleaner to have 1 CIP for the framework and 1 CIP for each new set of related keys. We were kind of stuck with CIP-0068 because we had to represent the state of the art until that point... but here we can start fresh, so there's no reason the framework CIP would have to include a "base" set of keys (I'd be happy to leave that to @stevenj this time).

I agree!

File: CIP-xxxx/README.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+Implementors: []
+Discussions:
+    - https://github.com/cardano-foundation/cips/pulls/753
+Created: 2024-1-19
+License: CC-BY-4.0
+---
+
+<!-- cspell: words secp -->
+
+<!-- markdownlint-disable MD025-->
```
Comment (Crypto2099) [inline CIP-xxxx/README.md:17]:
```suggestion
```

File: CIP-xxxx/README.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+# [CBOR] Tag definition for [CIP-0003] [ED25519-BIP32] Keys
```
Comment (Crypto2099) [inline CIP-xxxx/README.md:18]: @rphair correct me if I'm wrong, I believe we do not generally include an H1 tag as this is parsed from the yaml header in downstream renderers.

File: CIP-xxxx/README.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+## Abstract
+
+[CIP-0003] defines [ED25519-BIP32] Keys, Key derivation and a signature scheme.
+This CIP defines CBOR Tags and formalizes the Canonical encoding of data structures for [CIP-0003].
+The intention is to have these tags registered in the [IANA CBOR Tag Registry].
```
Comment (Crypto2099) [inline CIP-xxxx/README.md:24]:
```suggestion
The intention is to have these tags registered in the [IANA CBOR Tag Registry].
```

Is this explicitly a motivation of this CIP or a side-effect of the registry definition CIP?

File: CIP-xxxx/README.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+An official published spec is required to register these Tags with [IANA][IANA CBOR Tag Registry].
+This document also serves that purpose.
+
+## Path to Active
+
+### Acceptance Criteria
+
+* At least 1 Metadata CIP uses at least 1 of the tags defined in this CIP.
+* IANA register the Tags as defined.
```
Comment (Crypto2099) [inline CIP-xxxx/README.md:122]:
```suggestion
[ ] At least 1 Metadata CIP uses at least 1 of the tags defined in this CIP.
[ ] IANA register the Tags as defined.
```

File: CIP-xxxx/README.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+[Apache 2.0]: https://www.apache.org/licenses/LICENSE-2.0.html
+[CBOR]: https://www.rfc-editor.org/rfc/rfc8949.html
+[CDDL]: https://www.rfc-editor.org/rfc/rfc8610
+[CIP-0003]: https://cips.cardano.org/cip/CIP-0003
+[BIP32]: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
+[IANA CBOR Tag Registry]: https://www.iana.org/assignments/cbor-tags/cbor-tags.xhtml
+[network byte order]: https://datatracker.ietf.org/doc/html/rfc1700
+[ED25519-BIP32]: https://github.com/input-output-hk/adrestia/raw/bdf00e4e7791d610d273d227be877bc6dd0dbcfb/user-guide/static/Ed25519_BIP.pdf
+[CPS-CBOR-TAGS]: https://github.com/cardano-foundation/CIPs/pull/751
+[CIP-CBOR-TAGS]: https://github.com/cardano-foundation/CIPs/pull/752
```
Comment (Crypto2099) [inline CIP-xxxx/README.md:145]: @rphair can you provide context on recommendation per the new update to accept a **references** section in the standard template

**Review by Crypto2099 — COMMENTED** [review]
Minor personal preference but I would love to see a stripped down template of this CIP as a demonstration for future Tag definition CIPs.

**Comment by Crypto2099** [comment]
Going to echo some of the sentiment from @rphair and @Ryun1. I love how these three PRs have been drafted to work in unison and provide a "clean" template for how these "registry" types of CIPs/CPSs could be defined in the future. This seems much cleaner than what we have learned through past experience w/ CIP-10, 68, etc.

File: CIP-xxxx/README.md (hunk 2 above)
Comment (rphair) [reply CIP-xxxx/README.md:18]: @Crypto2099 you are right; this is something we tell everyone who uses an H1 tag & it's in CIP-0001 here (see **Note**): https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#structure

File: CIP-xxxx/README.md (hunk 5 above)
Comment (rphair) [reply CIP-xxxx/README.md:145]: @Crypto2099 yes... @stevenj I think our readers would benefit for creating markup for these, all given a heading of `## References` (and moving it before the _Copyright_) as recently endorsed here (see _optional sections_): https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#structure

File: CIP-xxxx/README.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
+---
+CIP: xxxx
+Title: CBOR Tag definition for CIP-0003 ED25519-BIP32 Keys
```
[the remark this answers is not in the document]
Comment (rphair) [reply CIP-xxxx/README.md:3]: Yes, please @stevenj - I'll make this change in the PR Title for consistency with our other CIP groups like `Web-Wallet Bridge` which you can see in the CIP index at this point.  Would be very helpful to have this as well for your & future CBOR registry extensions.  cc @Ryun1

File: CIP-xxxx/README.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+Category: MetaData
```
Comment (rphair) [inline CIP-xxxx/README.md:4]:
```suggestion
Category: Tools
```

**Review by rphair — COMMENTED** [review]
@stevenj the question posted in https://github.com/cardano-foundation/CIPs/pull/753#issuecomment-1912144155 has been answered:

This week's CIP meeting concluded by undisputed consensus that the CBOR Registry related CIPs and CPS are of `Category:` = `Tools` (rather than `Meta` or `Metadata`) in short because these formats will inevitably be consumed off-chain by a currently unknowable variety of applications.

File: CIP-xxxx/README.md (hunk 1 above)
Comment (stevenj) [reply CIP-xxxx/README.md]: Removed

File: CIP-xxxx/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (stevenj) [reply CIP-xxxx/README.md]: removed

File: CIP-xxxx/README.md (hunk 4 above)
Comment (stevenj) [reply CIP-xxxx/README.md]: Incorporated suggestion

File: CIP-xxxx/README.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
+---
+CIP: 0115
```
Comment (Ryun1) [inline CIP-xxxx/README.md:2]:
```suggestion
CIP: 115
```
no leading 0s for the header, I believe

File: CIP-xxxx/README.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+Title: CBOR tag definition - ED25519-BIP32 Keys
+Category: Tools
+Status: Proposed
+Authors:
+    - Steven Johnson<steven.johnson@iohk.io>
```
Comment (Ryun1) [inline CIP-xxxx/README.md:7]:
```suggestion
    - Steven Johnson <steven.johnson@iohk.io>
```
missing space

File: CIP-xxxx/README.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+Implementors: []
+Discussions:
+    - https://github.com/cardano-foundation/cips/pulls/753
+Created: 2024-1-19
```
Comment (Ryun1) [inline CIP-xxxx/README.md:11]:
```suggestion
Created: 2024-01-19
```

File: CIP-xxxx/README.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
+---
+CIP: 0115
+Title: CBOR tag definition - ED25519-BIP32 Keys
+Category: Tools
+Status: Proposed
+Authors:
+    - Steven Johnson<steven.johnson@iohk.io>
+Implementors: []
```
Comment (Ryun1) [inline CIP-xxxx/README.md:8]:
```suggestion
Implementors: N/A
```
Not sure we want to explicitly list things here?
or do we want to list tooling which does use this tag?

File: CIP-xxxx/README.md (hunk 3 above)
Comment (Ryun1) [reply CIP-xxxx/README.md:19]: I believe this is an explicit motivation

File: CIP-xxxx/README.md (hunk 5 above)
Comment (stevenj) [reply CIP-xxxx/README.md:153]: Done

File: CIP-xxxx/README.md (hunk 3 above)
Comment (stevenj) [reply CIP-xxxx/README.md:19]: Yes its explicit motivation.  You need a published source of the proposed tags to get IANA to register them.

Comment (rphair) [reply CIP-xxxx/README.md]: @stevenj @Ryun1 @Crypto2099 as I understand it then the matter is resolved (please unresolve if not)

File: CIP-0115/README.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+[CC-BY-4.0]: https://creativecommons.org/licenses/by/4.0/legalcode
+[Apache 2.0]: https://www.apache.org/licenses/LICENSE-2.0.html
+[CBOR]: https://www.rfc-editor.org/rfc/rfc8949.html
+[CDDL]: https://www.rfc-editor.org/rfc/rfc8610
+[CIP-0003]: https://cips.cardano.org/cip/CIP-0003
+[BIP32]: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
+[IANA CBOR Tag Registry]: https://www.iana.org/assignments/cbor-tags/cbor-tags.xhtml
+[network byte order]: https://datatracker.ietf.org/doc/html/rfc1700
+[ED25519-BIP32]: https://github.com/input-output-hk/adrestia/raw/bdf00e4e7791d610d273d227be877bc6dd0dbcfb/user-guide/static/Ed25519_BIP.pdf
+[CPS-CBOR-TAGS]: https://github.com/cardano-foundation/CIPs/pull/751
```
Comment (Ryun1) [inline CIP-0115/README.md:152]:
```suggestion
[CPS-0014]: https://github.com/cardano-foundation/CIPs/tree/master/CPS-0014
```

File: CIP-0115/README.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+This CIP defines CBOR Tags and formalizes the Canonical encoding of data structures for [CIP-0003].
+The intention is to have these tags registered in the [IANA CBOR Tag Registry].
+
+## Motivation: why is this CIP necessary?
+
+Project Catalyst is in the process of defining new CBOR data structures.
+We needed a way to reliably disambiguate different 32 byte strings.
+Rather than making a non-standard encoding scheme specific to our structures we would like to use standard [CBOR] Tags.
+
+This CIP is informed by [CPS-CBOR-TAGS] and [CIP-CBOR-TAGS].
```
Comment (Ryun1) [inline CIP-0115/README.md:27]:
```suggestion
This CIP is informed by [CPS-0014] and [CIP-0114].
```

File: CIP-0115/README.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+- [ ] IANA register all the tags as defined herein.
+
+### Implementation Plan
+
+- [ ] Tags are to be used by Project Catalyst for CBOR data structure definitions.
+- [ ] Project Catalyst will also make the application to [IANA][IANA CBOR Tag Registry] to register the Tags.
+
+## References
+
+- [CPS-0014 -  Register of CBOR Tags for Cardano Data structures][CPS-CBOR-TAGS]
```
Comment (Ryun1) [inline CIP-0115/README.md:127]:
```suggestion
- [CPS-0014 -  Register of CBOR Tags for Cardano Data structures][CPS-0014]
```

**Review by rphair — APPROVED** [review]
By my reckoning this now includes @Ryun1's last few spotted improvements & the admonitions (in advance of IANA official assignment of tags) @stevenj was talking about adding at today's meeting.  So I believe we are now all set to merge between meetings if other editors approve.

**Review by Ryun1 — APPROVED** [review]
good to go!

**Pull request merged**